### PR TITLE
4.x view custom options

### PR DESCRIPTION
**Pull request by ADmad**
Refs #13272

### CODE REVIEW

File: src/View/SerializedView.php [hunk 1]
```diff
@@ -107,4 +135,30 @@ public function render(?string $template = null, $layout = null): string
...
+        if (isset($this->{$name})) {
+            return $this->{$name};
+        }
+
+        if (isset($this->viewVars["_{$name}"])) {
+            deprecationWarning(sprintf(
+                'Setting special view var "_%s" is deprecated. Use ViewBuilder::setOption(\'%s\', $value) instead.',
+                $name,
+                $name
+            ));
```
Comment (markstory) [inline src/View/SerializedView.php]: I like this :+1:

File: src/View/XmlView.php [hunk 2]
```diff
@@ -80,33 +80,44 @@ class XmlView extends SerializedView
...
+
     /**
      * List of special view vars.
      *
+     * Use ViewBuilder::setOption()/setOptions() to set these vars.
+     *
      * @var array
      */
-    protected $_specialVars = ['_serialize', '_rootNode', '_xmlOptions'];
+    protected $_specialVars = ['serialize', 'rootNode', 'xmlOptions'];
```
Comment (markstory) [inline src/View/XmlView.php]: Is specialVars the best name for this? Would `context` or `options` be better?

**Comment by ADmad** [comment]
@markstory I removed the `$_specialVars` property and instead added `$options` for `View` itself which holds hash of custom view options. Having separate property for each custom option wasn't necessary. The `$options` hash will allow managing view option for user defined classes too.

But after implementing this I realized I partly re-implemented what `InstanceConfigTrait` provides :grimacing:. So should I make `View` just use `InstanceConfigTrait` instead?

File: src/View/JsonView.php [hunk 3]
```diff
@@ -79,33 +79,27 @@ class JsonView extends SerializedView
...
+     *   names. If true all view variables will be serialized. If null or false
+     *   normal view template will be rendered.
+     * - `jsonOptions`: Options for json_encode(). For e.g. `JSON_HEX_TAG | JSON_HEX_APOS`.
+     * - `jsonp`: Enables JSONP support and wraps response in callback function provided in query string.
+     *   - Setting it to true enables the default query string parameter "callback".
+     *   - Setting it to a string value, uses the provided query string parameter
      *     for finding the JSONP callback name.
      *
-     * @var true|string|null
+     * @var array{serialize:string|bool|null, jsonOptions: int|null, jsonp: bool|string|null}
```
Comment (markstory) [inline src/View/JsonView.php]: Is this something new in psalm?

Comment (ADmad) [reply src/View/JsonView.php]: Not new, just something we haven't used before. We should try to use it wherever dealing with arrays with fixed keys. https://psalm.dev/docs/docblock_type_syntax/#object-like-arrays

**Comment by ADmad** [comment]
I made an alternate branch making `View` use `InstanceConfigTrait`:
https://github.com/cakephp/cakephp/compare/4.x...ADmad:4.x-view-config

**Comment by markstory** [comment]
I like the InstanceConfig approach as it reuses concepts we use everywhere else in the framework.

**Comment by ADmad** [comment]
Closing in favor of #13301

**Pull request closed without merging**